streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollector.java

```diff
@@ -76,18 +81,30 @@ public RecordCollector(Producer<byte[], byte[]> producer, String streamTaskId) {
                 new ProducerRecord<>(record.topic(), partition, record.timestamp(), keyBytes, valBytes);
         final String topic = serializedRecord.topic();
 
-        this.producer.send(serializedRecord, new Callback() {
-            @Override
-            public void onCompletion(RecordMetadata metadata, Exception exception) {
-                if (exception == null) {
-                    TopicPartition tp = new TopicPartition(metadata.topic(), metadata.partition());
-                    offsets.put(tp, metadata.offset());
-                } else {
-                    String prefix = String.format("task [%s]", streamTaskId);
-                    log.error(String.format("%s Error sending record to topic %s", prefix, topic), exception);
+        for (int attempt = 1; attempt <= MAX_SEND_ATTEMPTS; attempt++) {
+            try {
+                this.producer.send(serializedRecord, new Callback() {
+                    @Override
+                    public void onCompletion(RecordMetadata metadata, Exception exception) {
+                        if (exception == null) {
+                            TopicPartition tp = new TopicPartition(metadata.topic(), metadata.partition());
+                            offsets.put(tp, metadata.offset());
+                        } else {
+                            String prefix = String.format("task [%s]", streamTaskId);
+                            log.error(String.format("%s Error sending record to topic %s", prefix, topic), exception);
+                        }
+                    }
+                });
+                return;
+            } catch (TimeoutException e) {
+                if (attempt == MAX_SEND_ATTEMPTS) {
+                    throw new StreamsException(String.format("failed to send record to topic %s after %d attempts", topic, attempt));
```

Comment:
Add the stream task id prefix here as well for both exception message and the warning log entry.